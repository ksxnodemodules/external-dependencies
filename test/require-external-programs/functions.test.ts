import path from 'path'
import { readFile } from 'fs-extra'

import libDefaultExport, {
  fromAllManifests,
  allManifests,
  classifyManifestList,
  InvalidManifestError,
  KEYWORD,
  KEYNAME,
  ManifestContentList
} from 'require-external-programs-lib'

import utilsDefaultExport, {
  groupByManifest,
  groupByProgram,
  examine
} from 'require-external-programs-utils'

import createVirtualEnvironment from './virtual-env'
const { apply } = createVirtualEnvironment()

import * as whichMockData from './__mocks__/which/data'

describe('require-external-programs-lib', () => {
  describe('default export', () => {
    it('is fromAllManifests', () => {
      expect(libDefaultExport).toBe(fromAllManifests)
    })
  })

  describe('fromAllManifests', () => {
    const getMap = async (...paths: string[]) =>
      new Map(await fromAllManifests(path.join(...paths)))

    it('works with valid data', apply(async () => {
      expect(await getMap('valid')).toMatchSnapshot()
    }))

    describe('throws error when received invalid data', () => {
      const mkfn = (...paths: string[]) => apply(async () => {
        await getMap(...paths).then(
          resolve => {
            console.error({ resolve })
            throw new Error('Expecting the promise to reject, but it resolves')
          },
          error => {
            if (error instanceof InvalidManifestError) {
              expect({
                error,
                map: new Map(error.map)
              }).toMatchSnapshot()

              return
            }

            throw new Error(`Expecting an InvalidManifestError but received ${error}`)
          }
        )
      })

      it(`whose "peerDependencies" lacks "${KEYWORD}"`, mkfn('invalid', 'missing-dependency-in-peer'))
      it('which lacks "peerDependencies"', mkfn('invalid', 'missing-field-of-peer'))
      it(`whose "${KEYNAME}" is not an array`, mkfn('invalid', 'not-an-array'))
      it('which is located in /invalid', mkfn('invalid'))
      it('which is located in /', mkfn())
    })
  })

  describe('allManifests', () => {
    it('lists all manifests', apply(async () => {
      expect(await allManifests('.')).toMatchSnapshot()
    }))
  })

  describe('classifyManifestList', () => {
    it('matches snapshot', apply(async () => {
      const files = await allManifests('.')

      const list: ManifestContentList = await Promise.all(
        files.map(async manifestPath => ({
          manifestPath,
          manifestContent: JSON.parse(await readFile(manifestPath, { encoding: 'utf8' }))
        }))
      )

      expect(classifyManifestList(list)).toMatchSnapshot()
    }))
  })
})

describe('require-external-programs-utils', () => {
  describe('default export', () => {
    expect(utilsDefaultExport).toBe(examine)
  })

  describe('examine', () => {
    it('when all is satisfied', apply(async () => {
      const { getParams } = whichMockData.mockSync(cmd => `/bin/${cmd}`)
      const message = await examine('valid')
      const params = getParams()
      whichMockData.restoreAll()

      expect(message).toBe(false)
      expect(params).toMatchSnapshot()
    }))

    it('when all is not satisfied', apply(async () => {
      const { getParams } = whichMockData.mockSync(() => null)
      const message = await examine('valid')
      const params = getParams()
      whichMockData.restoreAll()

      expect(message).not.toBe(false)
      expect(message).toMatchSnapshot()
      expect(params).toMatchSnapshot()
    }))

    it('when only some is satisfied', apply(async () => {
      const availables = ['abc', 'def']

      const { getParams } = whichMockData
        .mockSync(cmd => availables.includes(cmd) ? `/bin/${cmd}` : null)

      const message = await examine('valid')
      const params = getParams()
      whichMockData.restoreAll()

      expect(message).not.toBe(false)
      expect(message).toMatchSnapshot()
      expect(params).toMatchSnapshot()
    }))
  })

  describe('group', () => {
    const item =
      (manifest: string, program: string) =>
        ({ manifest, program })

    const collection = new Set([
      item('foo/package.json', 'abc'),
      item('foo/package.json', 'def'),
      item('foo/package.json', 'ghi'),
      item('foo/package.json', 'abc'),
      item('bar/package.json', 'abc'),
      item('bar/package.json', 'jkl'),
      item('baz/package.json', 'mno')
    ])

    it('by manifests', () => {
      expect(new Map(groupByManifest(collection))).toMatchSnapshot()
    })

    it('by program', () => {
      expect(new Map(groupByProgram(collection))).toMatchSnapshot()
    })
  })
})
