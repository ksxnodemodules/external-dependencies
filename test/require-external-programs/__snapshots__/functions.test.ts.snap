// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`require-external-programs-lib allManifests lists all manifests 1`] = `
Array [
  "invalid/found-dependency-in-peer/qux/package.json",
  "invalid/lack-dependency-dicts/bar/package.json",
  "invalid/missing-field-in-dependency-dict/foo/package.json",
  "invalid/not-an-array/baz/package.json",
  "valid/node_modules/abc-def/package.json",
  "valid/node_modules/abc-ghi/package.json",
  "valid/node_modules/jkl/package.json",
  "valid/node_modules/mismatched-version/package.json",
  "valid/node_modules/missing-activation-key/package.json",
]
`;

exports[`require-external-programs-lib classifyManifestList matches snapshot 1`] = `
Object {
  "excluded": Array [
    Object {
      "manifestContent": Object {
        "dependencies": Object {
          "require-external-programs": "> 999 < 1",
        },
        "name": "mismatched-version",
        "requireExternalPrograms": Array [
          "hello",
        ],
      },
      "manifestPath": "valid/node_modules/mismatched-version/package.json",
    },
    Object {
      "manifestContent": Object {
        "devDependencies": Object {
          "require-external-programs": "*",
        },
        "name": "missing-activation-key",
      },
      "manifestPath": "valid/node_modules/missing-activation-key/package.json",
    },
  ],
  "included": Array [
    Object {
      "manifestContent": Object {
        "dependencies": Object {
          "require-external-programs": "*",
        },
        "name": "abc-def",
        "requireExternalPrograms": Array [
          "abc",
          "def",
        ],
      },
      "manifestPath": "valid/node_modules/abc-def/package.json",
    },
    Object {
      "manifestContent": Object {
        "devDependencies": Object {
          "require-external-programs": "*",
        },
        "name": "abc-ghi",
        "requireExternalPrograms": Array [
          "abc",
          "ghi",
        ],
      },
      "manifestPath": "valid/node_modules/abc-ghi/package.json",
    },
    Object {
      "manifestContent": Object {
        "dependencies": Object {
          "require-external-programs": "*",
        },
        "name": "jkl",
        "requireExternalPrograms": Array [
          "jkl",
        ],
      },
      "manifestPath": "valid/node_modules/jkl/package.json",
    },
  ],
  "invalid": Array [
    Object {
      "manifestContent": Object {
        "dependencies": Object {
          "require-external-programs": "*",
        },
        "name": "qux",
        "peerDependencies": Object {
          "require-external-programs": "*",
        },
      },
      "manifestPath": "invalid/found-dependency-in-peer/qux/package.json",
      "reason": "Found \\"require-external-programs\\" in \\"peerDependencies\\", use \\"dependencies\\" instead",
    },
    Object {
      "manifestContent": Object {
        "name": "bar",
        "requireExternalPrograms": Array [
          "hello",
        ],
      },
      "manifestPath": "invalid/lack-dependency-dicts/bar/package.json",
      "reason": "Found \\"requireExternalPrograms\\" but not \\"require-external-programs\\" in either \\"dependencies\\" or \\"devDependencies\\"",
    },
    Object {
      "manifestContent": Object {
        "dependencies": Object {},
        "devDependencies": Object {},
        "name": "foo",
        "requireExternalPrograms": Array [
          "hello",
        ],
      },
      "manifestPath": "invalid/missing-field-in-dependency-dict/foo/package.json",
      "reason": "Found \\"requireExternalPrograms\\" but not \\"require-external-programs\\" in either \\"dependencies\\" or \\"devDependencies\\"",
    },
    Object {
      "manifestContent": Object {
        "dependencies": Object {
          "require-external-programs": "*",
        },
        "name": "bar",
        "requireExternalPrograms": "hello",
      },
      "manifestPath": "invalid/not-an-array/baz/package.json",
      "reason": "Expecting \\"require-external-programs\\" to be an array but received \\"hello\\"",
    },
  ],
}
`;

exports[`require-external-programs-lib fromAllManifests throws error when received invalid data which is located in / 1`] = `
Object {
  "error": [InvalidManifest: Detected 4 invalid items],
  "map": Map {
    "invalid/found-dependency-in-peer/qux/package.json" => Object {
      "manifestContent": Object {
        "dependencies": Object {
          "require-external-programs": "*",
        },
        "name": "qux",
        "peerDependencies": Object {
          "require-external-programs": "*",
        },
      },
      "manifestPath": "invalid/found-dependency-in-peer/qux/package.json",
      "reason": "Found \\"require-external-programs\\" in \\"peerDependencies\\", use \\"dependencies\\" instead",
    },
    "invalid/lack-dependency-dicts/bar/package.json" => Object {
      "manifestContent": Object {
        "name": "bar",
        "requireExternalPrograms": Array [
          "hello",
        ],
      },
      "manifestPath": "invalid/lack-dependency-dicts/bar/package.json",
      "reason": "Found \\"requireExternalPrograms\\" but not \\"require-external-programs\\" in either \\"dependencies\\" or \\"devDependencies\\"",
    },
    "invalid/missing-field-in-dependency-dict/foo/package.json" => Object {
      "manifestContent": Object {
        "dependencies": Object {},
        "devDependencies": Object {},
        "name": "foo",
        "requireExternalPrograms": Array [
          "hello",
        ],
      },
      "manifestPath": "invalid/missing-field-in-dependency-dict/foo/package.json",
      "reason": "Found \\"requireExternalPrograms\\" but not \\"require-external-programs\\" in either \\"dependencies\\" or \\"devDependencies\\"",
    },
    "invalid/not-an-array/baz/package.json" => Object {
      "manifestContent": Object {
        "dependencies": Object {
          "require-external-programs": "*",
        },
        "name": "bar",
        "requireExternalPrograms": "hello",
      },
      "manifestPath": "invalid/not-an-array/baz/package.json",
      "reason": "Expecting \\"require-external-programs\\" to be an array but received \\"hello\\"",
    },
  },
}
`;

exports[`require-external-programs-lib fromAllManifests throws error when received invalid data which is located in /invalid 1`] = `
Object {
  "error": [InvalidManifest: Detected 4 invalid items],
  "map": Map {
    "invalid/found-dependency-in-peer/qux/package.json" => Object {
      "manifestContent": Object {
        "dependencies": Object {
          "require-external-programs": "*",
        },
        "name": "qux",
        "peerDependencies": Object {
          "require-external-programs": "*",
        },
      },
      "manifestPath": "invalid/found-dependency-in-peer/qux/package.json",
      "reason": "Found \\"require-external-programs\\" in \\"peerDependencies\\", use \\"dependencies\\" instead",
    },
    "invalid/lack-dependency-dicts/bar/package.json" => Object {
      "manifestContent": Object {
        "name": "bar",
        "requireExternalPrograms": Array [
          "hello",
        ],
      },
      "manifestPath": "invalid/lack-dependency-dicts/bar/package.json",
      "reason": "Found \\"requireExternalPrograms\\" but not \\"require-external-programs\\" in either \\"dependencies\\" or \\"devDependencies\\"",
    },
    "invalid/missing-field-in-dependency-dict/foo/package.json" => Object {
      "manifestContent": Object {
        "dependencies": Object {},
        "devDependencies": Object {},
        "name": "foo",
        "requireExternalPrograms": Array [
          "hello",
        ],
      },
      "manifestPath": "invalid/missing-field-in-dependency-dict/foo/package.json",
      "reason": "Found \\"requireExternalPrograms\\" but not \\"require-external-programs\\" in either \\"dependencies\\" or \\"devDependencies\\"",
    },
    "invalid/not-an-array/baz/package.json" => Object {
      "manifestContent": Object {
        "dependencies": Object {
          "require-external-programs": "*",
        },
        "name": "bar",
        "requireExternalPrograms": "hello",
      },
      "manifestPath": "invalid/not-an-array/baz/package.json",
      "reason": "Expecting \\"require-external-programs\\" to be an array but received \\"hello\\"",
    },
  },
}
`;

exports[`require-external-programs-lib fromAllManifests throws error when received invalid data which lacks both "dependencies" and "devDependencies 1`] = `
Object {
  "error": [InvalidManifest: Detected 1 invalid items],
  "map": Map {
    "invalid/lack-dependency-dicts/bar/package.json" => Object {
      "manifestContent": Object {
        "name": "bar",
        "requireExternalPrograms": Array [
          "hello",
        ],
      },
      "manifestPath": "invalid/lack-dependency-dicts/bar/package.json",
      "reason": "Found \\"requireExternalPrograms\\" but not \\"require-external-programs\\" in either \\"dependencies\\" or \\"devDependencies\\"",
    },
  },
}
`;

exports[`require-external-programs-lib fromAllManifests throws error when received invalid data whose "dependencies" and "devDependencies" lack "require-external-programs" 1`] = `
Object {
  "error": [InvalidManifest: Detected 1 invalid items],
  "map": Map {
    "invalid/missing-field-in-dependency-dict/foo/package.json" => Object {
      "manifestContent": Object {
        "dependencies": Object {},
        "devDependencies": Object {},
        "name": "foo",
        "requireExternalPrograms": Array [
          "hello",
        ],
      },
      "manifestPath": "invalid/missing-field-in-dependency-dict/foo/package.json",
      "reason": "Found \\"requireExternalPrograms\\" but not \\"require-external-programs\\" in either \\"dependencies\\" or \\"devDependencies\\"",
    },
  },
}
`;

exports[`require-external-programs-lib fromAllManifests throws error when received invalid data whose "requireExternalPrograms" is not an array 1`] = `
Object {
  "error": [InvalidManifest: Detected 1 invalid items],
  "map": Map {
    "invalid/not-an-array/baz/package.json" => Object {
      "manifestContent": Object {
        "dependencies": Object {
          "require-external-programs": "*",
        },
        "name": "bar",
        "requireExternalPrograms": "hello",
      },
      "manifestPath": "invalid/not-an-array/baz/package.json",
      "reason": "Expecting \\"require-external-programs\\" to be an array but received \\"hello\\"",
    },
  },
}
`;

exports[`require-external-programs-lib fromAllManifests works with valid data 1`] = `
Map {
  "valid/node_modules/abc-def/package.json" => "def",
  "valid/node_modules/abc-ghi/package.json" => "ghi",
  "valid/node_modules/jkl/package.json" => "jkl",
}
`;

exports[`require-external-programs-utils examine when all is not satisfied 1`] = `
Object {
  "error": false,
  "response": "
Missing 4 external programs
[2m* [22m[1mabc[22m
[2m  - [22m[2m[3mrequired by [23m[22m[3mvalid/node_modules/abc-def/package.json[23m
[2m  - [22m[2m[3mrequired by [23m[22m[3mvalid/node_modules/abc-ghi/package.json[23m

[2m* [22m[1mdef[22m
[2m  - [22m[2m[3mrequired by [23m[22m[3mvalid/node_modules/abc-def/package.json[23m

[2m* [22m[1mghi[22m
[2m  - [22m[2m[3mrequired by [23m[22m[3mvalid/node_modules/abc-ghi/package.json[23m

[2m* [22m[1mjkl[22m
[2m  - [22m[2m[3mrequired by [23m[22m[3mvalid/node_modules/jkl/package.json[23m

",
  "satisfied": false,
}
`;

exports[`require-external-programs-utils examine when all is not satisfied 2`] = `
Set {
  Array [
    "abc",
    Object {
      "nothrow": true,
    },
  ],
  Array [
    "def",
    Object {
      "nothrow": true,
    },
  ],
  Array [
    "ghi",
    Object {
      "nothrow": true,
    },
  ],
  Array [
    "jkl",
    Object {
      "nothrow": true,
    },
  ],
}
`;

exports[`require-external-programs-utils examine when all is satisfied 1`] = `
Set {
  Array [
    "abc",
    Object {
      "nothrow": true,
    },
  ],
  Array [
    "def",
    Object {
      "nothrow": true,
    },
  ],
  Array [
    "ghi",
    Object {
      "nothrow": true,
    },
  ],
  Array [
    "jkl",
    Object {
      "nothrow": true,
    },
  ],
}
`;

exports[`require-external-programs-utils examine when only some is satisfied 1`] = `
Object {
  "error": false,
  "response": "
Missing 2 external programs
[2m* [22m[1mghi[22m
[2m  - [22m[2m[3mrequired by [23m[22m[3mvalid/node_modules/abc-ghi/package.json[23m

[2m* [22m[1mjkl[22m
[2m  - [22m[2m[3mrequired by [23m[22m[3mvalid/node_modules/jkl/package.json[23m

",
  "satisfied": false,
}
`;

exports[`require-external-programs-utils examine when only some is satisfied 2`] = `
Set {
  Array [
    "abc",
    Object {
      "nothrow": true,
    },
  ],
  Array [
    "def",
    Object {
      "nothrow": true,
    },
  ],
  Array [
    "ghi",
    Object {
      "nothrow": true,
    },
  ],
  Array [
    "jkl",
    Object {
      "nothrow": true,
    },
  ],
}
`;

exports[`require-external-programs-utils examine when some manifest files are invalid 1`] = `
Object {
  "error": true,
  "response": "
[[1mInvalidManifest:[22m Detected 4 invalid items]
  * Found \\"require-external-programs\\" in \\"peerDependencies\\", use \\"dependencies\\" instead [2m(invalid/found-dependency-in-peer/qux/package.json)[22m
  * Found \\"requireExternalPrograms\\" but not \\"require-external-programs\\" in either \\"dependencies\\" or \\"devDependencies\\" [2m(invalid/lack-dependency-dicts/bar/package.json)[22m
  * Found \\"requireExternalPrograms\\" but not \\"require-external-programs\\" in either \\"dependencies\\" or \\"devDependencies\\" [2m(invalid/missing-field-in-dependency-dict/foo/package.json)[22m
  * Expecting \\"require-external-programs\\" to be an array but received \\"hello\\" [2m(invalid/not-an-array/baz/package.json)[22m
",
  "satisfied": false,
}
`;

exports[`require-external-programs-utils execute when all is not satisfied 1`] = `
Object {
  "exitParams": Array [
    Array [
      1,
    ],
  ],
  "logParams": Array [
    Array [
      "
Missing 4 external programs
[2m* [22m[1mabc[22m
[2m  - [22m[2m[3mrequired by [23m[22m[3mvalid/node_modules/abc-def/package.json[23m
[2m  - [22m[2m[3mrequired by [23m[22m[3mvalid/node_modules/abc-ghi/package.json[23m

[2m* [22m[1mdef[22m
[2m  - [22m[2m[3mrequired by [23m[22m[3mvalid/node_modules/abc-def/package.json[23m

[2m* [22m[1mghi[22m
[2m  - [22m[2m[3mrequired by [23m[22m[3mvalid/node_modules/abc-ghi/package.json[23m

[2m* [22m[1mjkl[22m
[2m  - [22m[2m[3mrequired by [23m[22m[3mvalid/node_modules/jkl/package.json[23m

",
    ],
  ],
}
`;

exports[`require-external-programs-utils execute when all is satisfied 1`] = `
Object {
  "exitParams": Array [
    Array [
      0,
    ],
  ],
  "logParams": Array [],
}
`;

exports[`require-external-programs-utils execute when only some is satisfied 1`] = `
Object {
  "exitParams": Array [
    Array [
      1,
    ],
  ],
  "logParams": Array [
    Array [
      "
Missing 2 external programs
[2m* [22m[1mghi[22m
[2m  - [22m[2m[3mrequired by [23m[22m[3mvalid/node_modules/abc-ghi/package.json[23m

[2m* [22m[1mjkl[22m
[2m  - [22m[2m[3mrequired by [23m[22m[3mvalid/node_modules/jkl/package.json[23m

",
    ],
  ],
}
`;

exports[`require-external-programs-utils execute when some manifest files are invalid 1`] = `
Object {
  "exitParams": Array [
    Array [
      2,
    ],
  ],
  "logParams": Array [
    Array [
      "
[[1mInvalidManifest:[22m Detected 4 invalid items]
  * Found \\"require-external-programs\\" in \\"peerDependencies\\", use \\"dependencies\\" instead [2m(invalid/found-dependency-in-peer/qux/package.json)[22m
  * Found \\"requireExternalPrograms\\" but not \\"require-external-programs\\" in either \\"dependencies\\" or \\"devDependencies\\" [2m(invalid/lack-dependency-dicts/bar/package.json)[22m
  * Found \\"requireExternalPrograms\\" but not \\"require-external-programs\\" in either \\"dependencies\\" or \\"devDependencies\\" [2m(invalid/missing-field-in-dependency-dict/foo/package.json)[22m
  * Expecting \\"require-external-programs\\" to be an array but received \\"hello\\" [2m(invalid/not-an-array/baz/package.json)[22m
",
    ],
  ],
}
`;

exports[`require-external-programs-utils group by manifests 1`] = `
Map {
  "foo/package.json" => Set {
    "abc",
    "def",
    "ghi",
  },
  "bar/package.json" => Set {
    "abc",
    "jkl",
  },
  "baz/package.json" => Set {
    "mno",
  },
}
`;

exports[`require-external-programs-utils group by program 1`] = `
Map {
  "abc" => Set {
    "foo/package.json",
    "bar/package.json",
  },
  "def" => Set {
    "foo/package.json",
  },
  "ghi" => Set {
    "foo/package.json",
  },
  "jkl" => Set {
    "bar/package.json",
  },
  "mno" => Set {
    "baz/package.json",
  },
}
`;

exports[`require-external-programs-utils messageFromError when error is an instance of Error 1`] = `[Error: Hello, World!!]`;

exports[`require-external-programs-utils messageFromError when error is an instance of InvalidManifest 1`] = `
"
[[1mInvalidManifest:[22m Detected 4 invalid items]
  * Found \\"require-external-programs\\" in \\"peerDependencies\\", use \\"dependencies\\" instead [2m(invalid/found-dependency-in-peer/qux/package.json)[22m
  * Found \\"requireExternalPrograms\\" but not \\"require-external-programs\\" in either \\"dependencies\\" or \\"devDependencies\\" [2m(invalid/lack-dependency-dicts/bar/package.json)[22m
  * Found \\"requireExternalPrograms\\" but not \\"require-external-programs\\" in either \\"dependencies\\" or \\"devDependencies\\" [2m(invalid/missing-field-in-dependency-dict/foo/package.json)[22m
  * Expecting \\"require-external-programs\\" to be an array but received \\"hello\\" [2m(invalid/not-an-array/baz/package.json)[22m
"
`;
